webknossos/webknossos/dataset/dataset.py

```diff
@@ -750,6 +755,118 @@ def add_layer_for_existing_files(
+        self,
+        images: Union[str, "pims.FramesSequence", List[Union[str, PathLike]]],
+        ## add_layer arguments
+        layer_name: str,
+        category: LayerCategoryType = "color",
+        data_format: Union[str, DataFormat] = DEFAULT_DATA_FORMAT,
+        ## add_mag arguments
+        mag: Union[int, str, list, tuple, np.ndarray, Mag] = Mag(1),
+        chunk_size: Optional[Union[Vec3IntLike, int]] = None,
+        chunks_per_shard: Optional[Union[int, Vec3IntLike]] = None,
+        compress: bool = False,
+        ## other arguments
+        swap_xy: bool = False,
+        flip_x: bool = False,
+        flip_y: bool = False,
+        flip_z: bool = False,
+        use_bioformats: bool = False,
+        timepoint: Optional[int] = None,
+        batch_size: Optional[int] = None,  # defaults to shard-size z
+        executor: Optional[Union[ClusterExecutor, WrappedProcessPoolExecutor]] = None,
+    ) -> Layer:
+        try:
+            from ._utils.pims_images import PimsImages, dimwise_max
+        except ImportError as e:
+            raise RuntimeError(
+                "Cannot import pims, please install e.g. using 'webknossos[all]'"
+            ) from e
+
+        pims_images = PimsImages(
+            images,
+            timepoint=timepoint,
+            swap_xy=swap_xy,
+            flip_x=flip_x,
+            flip_y=flip_y,
+            flip_z=flip_z,
+            use_bioformats=use_bioformats,
+        )
+
+        layer = self.add_layer(
+            layer_name=layer_name,
+            category=category,
+            data_format=data_format,
+            dtype_per_channel=pims_images.dtype,
+            num_channels=pims_images.num_channels,
+        )
+        mag_view = layer.add_mag(
+            mag=mag,
+            chunk_size=chunk_size,
+            chunks_per_shard=chunks_per_shard,
+            compress=compress,
+        )
+        mag = mag_view.mag
+        layer.bounding_box = BoundingBox(
+            (0, 0, 0), pims_images.expected_shape
+        ).from_mag_to_mag1(mag)
+
+        if batch_size is None:
+            batch_size = mag_view.info.shard_size.z
+        else:
+            assert (
+                batch_size % mag_view.info.shard_size.z == 0
+            ), f"batch_size {batch_size} must be divisible by z shard-size {mag_view.info.shard_size.z}"
+
+        func_per_chunk = partial(
+            pims_images.copy_to_view,
+            mag_view=mag_view,
+        )
+
+        args = []
+        for z_start in range(0, pims_images.expected_shape.z, batch_size):
+            z_end = min(z_start + batch_size, pims_images.expected_shape.z)
+            # return shapes and set to union when using --pad
+            args.append((z_start, z_end))
+        with warnings.catch_warnings():
+            # if a slice is larger than the others it might happen that
+            # a new chunk is partially written, leading to a warning,
+            # which is ignored in this context
```

Comment:
I think in the first case this should work, since the first write will set the new bounding-box, which then fits all subsequent writes. Writes must be either shard or bounding-box aligned (per bbox-border). I think this warning is not useful at all in this case, since it only happens at the dataset-borders, where the user can not change much. The only important warning is about the different sizes, which is handled elsewhere. The performance penalty should be negligible, since it's only about the borders of the dataset. It's not that the general chunk-size doesn't fit the blocks during iteration.

Reply:
I expanded the comment, I hope it makes more sense now :+1: